### PR TITLE
Add in-memory transpose for pull algorithms and refactor Gluon

### PR DESCRIPTION
**Pull request by gurbinder533**

### CODE REVIEW

File: libgluon/include/galois/graphs/GluonSubstrate.h [hunk 1]
```diff
@@ -2844,10 +2825,12 @@ class GluonSubstrate : public galois::runtime::GlobalObject {
   inline void sync_src_to_any(std::string loopName) {
     // only broadcast for OEC
     // reduce and broadcast for IEC, CVC, UVC
-    if (transposed || isVertexCut) {
+    if (!is_outgoing_cut_) {
       reduce<writeSource, readAny, SyncFnTy, BitsetFnTy, async>(loopName);
+      broadcast<writeSource, readAny, SyncFnTy, BitsetFnTy, async>(loopName);
+    } else {
+      broadcast<writeSource, readAny, SyncFnTy, BitsetFnTy, async>(loopName);
```
Comment (roshandathathri) [inline libgluon/include/galois/graphs/GluonSubstrate.h:2832]: This is the same for if and else, so move it out.

File: libgluon/include/galois/graphs/GluonSubstrate.h [hunk 2]
```diff
@@ -2863,20 +2846,13 @@ class GluonSubstrate : public galois::runtime::GlobalObject {
...
-      if (isVertexCut) {
-        reduce<writeDestination, readSource, SyncFnTy, BitsetFnTy, async>(
-            loopName);
-      }
+    if (!is_outgoing_cut_) {
       broadcast<writeDestination, readSource, SyncFnTy, BitsetFnTy, async>(
           loopName);
-    } else {
+    }
+    if (!is_incoming_cut_) {
```
Comment (roshandathathri) [inline libgluon/include/galois/graphs/GluonSubstrate.h:2853]: Move this if block before the previous if block. Always, reduce first and then broadcast.

File: libgluon/include/galois/graphs/GluonSubstrate.h [hunk 3]
```diff
@@ -2912,10 +2888,15 @@ class GluonSubstrate : public galois::runtime::GlobalObject {
...
     // reduce and broadcast for OEC, CVC, UVC
-    if (!transposed || isVertexCut) {
-      reduce<writeDestination, readAny, SyncFnTy, BitsetFnTy, async>(loopName);
+    if (!is_incoming_cut_) {
+      reduce<writeDestination, readDestination, SyncFnTy, BitsetFnTy, async>(
+          loopName);
+      broadcast<writeDestination, readDestination, SyncFnTy, BitsetFnTy, async>(
+          loopName);
+    } else {
+      broadcast<writeDestination, readDestination, SyncFnTy, BitsetFnTy, async>(
```
Comment (roshandathathri) [inline libgluon/include/galois/graphs/GluonSubstrate.h:2897]: Common for both if and else block, so move it out.

File: libgluon/include/galois/graphs/GluonSubstrate.h [hunk 4]
```diff
@@ -2912,10 +2888,15 @@ class GluonSubstrate : public galois::runtime::GlobalObject {
   inline void sync_dst_to_any(std::string loopName) {
     // only broadcast for IEC
     // reduce and broadcast for OEC, CVC, UVC
-    if (!transposed || isVertexCut) {
-      reduce<writeDestination, readAny, SyncFnTy, BitsetFnTy, async>(loopName);
+    if (!is_incoming_cut_) {
+      reduce<writeDestination, readDestination, SyncFnTy, BitsetFnTy, async>(
```
Comment (roshandathathri) [inline libgluon/include/galois/graphs/GluonSubstrate.h:2892]: Seems to be a typo. It's readAny for all functions within dst_to_any.

File: libgluon/include/galois/graphs/GluonSubstrate.h [hunk 5]
```diff
@@ -2997,32 +2978,60 @@ class GluonSubstrate : public galois::runtime::GlobalObject {
...
-          sync_src_to_any<SyncFnTy, BitsetFnTy, async>(loopName);
-        }
-      } else if (writeLocation == writeDestination) {
-        if (readLocation == readSource) {
-          sync_dst_to_src<SyncFnTy, BitsetFnTy, async>(loopName);
-        } else if (readLocation == readDestination) {
-          sync_dst_to_dst<SyncFnTy, BitsetFnTy, async>(loopName);
-        } else { // readAny
-          sync_dst_to_any<SyncFnTy, BitsetFnTy, async>(loopName);
+      if (transposed_) {
```
Comment (roshandathathri) [inline libgluon/include/galois/graphs/GluonSubstrate.h:2984]: Add a comment saying "swap source and destination for read/write location (does not apply to any)"

**Review by roshandathathri — APPROVED** [review]
Yay! Thanks for doing this.

Except sync_dst_to_src(), rest LGTM. The other comments are minor nits.

File: libcusp/include/galois/graphs/Partitioner.h [hunk 6]
```diff
@@ -3269,6 +3269,91 @@ class Partitioner {
...
+
+    // Reuse out_indices_tmp for computing new destination positions
+    if (dist_.num_nodes_ >= 1) {
+      out_indices_tmp[0] = 0;
+      galois::do_all(
+          galois::iterate(static_cast<uint32_t>(1), dist_.num_nodes_),
+          [&](uint32_t n) {
+            out_indices_tmp[n] = dist_.large_array_out_indices_[n - 1];
+          },
+          galois::no_stats());
```
[the remark this answers is not in the document]
Comment (gurbinder533) [reply libcusp/include/galois/graphs/Partitioner.h:3328]: It is not exactly swapping actually.

[the remark this answers is not in the document]
Comment (roshandathathri) [reply libcusp/include/galois/graphs/Partitioner.h:3328]: I got that but the downstream code can be made to work with swapping if I understand right?

**Pull request merged**